controllers/kubeletconfig_controller.go

```diff
@@ -275,6 +295,51 @@ func (r *KubeletConfigReconciler) makeKCHandlerForPlatform(ctx context.Context,
 	return nil, fmt.Errorf("unsupported platform: %s", r.Platform)
 }
 
+func (r *KubeletConfigReconciler) deleteConfigMap(ctx context.Context, instance *nropv1.NUMAResourcesOperator, kcKey client.ObjectKey) (*corev1.ConfigMap, bool, error) {
+	cm := getDeletedOwner(kcKey, r.garbageCollectionFor)
+	if cm == nil {
+		return nil, false, nil
+	}
+	// we'll get to this flow only on hypershift
+	// on openshift the deletion is done automatically by setting the owner reference on the dependent ConfigMap
+	nodePoolName := cm.Labels[HyperShiftNodePoolLabel]
+	generatedName := objectnames.GetComponentName(instance.Name, nodePoolName)
+	dependentCM := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      generatedName,
+			Namespace: r.Namespace,
+		},
+	}
+	if err := r.Delete(ctx, dependentCM); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return cm, true, err
+		}
+		klog.V(2).InfoS("could not delete ConfigMap since it was not found", "configmapName", dependentCM.Name)
+	}
+	r.garbageCollectionFor = removeDeletedOwner(kcKey, r.garbageCollectionFor)
+	return cm, true, nil
+}
+
+func getDeletedOwner(kcKey client.ObjectKey, ownerConfigMaps []*corev1.ConfigMap) *corev1.ConfigMap {
+	for i := range ownerConfigMaps {
+		cm := ownerConfigMaps[i]
+		if client.ObjectKeyFromObject(cm).String() == kcKey.String() {
+			return cm
+		}
+	}
+	return nil
+}
+
+func removeDeletedOwner(kcKey client.ObjectKey, ownerConfigMaps []*corev1.ConfigMap) []*corev1.ConfigMap {
+	for i := range ownerConfigMaps {
+		cm := ownerConfigMaps[i]
+		if client.ObjectKeyFromObject(cm) == kcKey {
+			return append(ownerConfigMaps[:i], ownerConfigMaps[i+1:]...)
+		}
+	}
+	return ownerConfigMaps
+}
```

Comment:
This does not look obviously wrong but makes me wary. I'd go for the trivialest implementation:
```
func removeDeletedOwner(kcKey client.ObjectKey, ownerConfigMaps []*corev1.ConfigMap) []*corev1.ConfigMap {
        ret := make([]*corev1.ConfigMap, 0, len(ownerConfigMaps))
	for i := range ownerConfigMaps {
		cm := ownerConfigMaps[i]
		if client.ObjectKeyFromObject(cm) == kcKey {
			continue
		}
                ret = append(ret, cm)
	}
	return ret
}
```
perhaps something in the `slices` package can help?

Reply:
maybe https://pkg.go.dev/slices#DeleteFunc ?

Reply:
Nice. I used https://pkg.go.dev/slices#Delete which is good enough